core/src/main/java/io/questdb/cairo/wal/WalWriter.java

```diff
@@ -0,0 +1,1621 @@
+import io.questdb.cairo.vm.api.MemoryMAR;
+import io.questdb.cairo.vm.api.NullMemory;
+import io.questdb.cairo.wal.seq.*;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.engine.functions.constants.Long128Constant;
+import io.questdb.griffin.engine.ops.AbstractOperation;
+import io.questdb.griffin.engine.ops.AlterOperation;
+import io.questdb.griffin.engine.ops.UpdateOperation;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.*;
+import io.questdb.std.datetime.millitime.MillisecondClock;
+import io.questdb.std.str.Path;
+import io.questdb.std.str.SingleCharCharSequence;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.TestOnly;
+
+import static io.questdb.cairo.TableUtils.*;
+import static io.questdb.cairo.wal.seq.TableSequencer.NO_TXN;
+import static io.questdb.cairo.wal.WalUtils.WAL_NAME_BASE;
+
+public class WalWriter implements TableWriterAPI {
+    public static final int NEW_COL_RECORD_SIZE = 6;
+    private static final Log LOG = LogFactory.getLog(WalWriter.class);
+    private static final Runnable NOOP = () -> {
+    };
+    private static final int MEM_TAG = MemoryTag.MMAP_TABLE_WAL_WRITER;
+    private static final long COLUMN_DELETED_NULL_FLAG = Long.MAX_VALUE;
+    private final ObjList<MemoryMA> columns;
+    private final ObjList<SymbolMapReader> symbolMapReaders = new ObjList<>();
+    private final IntList initialSymbolCounts = new IntList();
+    private final ObjList<CharSequenceIntHashMap> symbolMaps = new ObjList<>();
+    private final MillisecondClock millisecondClock;
+    private final Path path;
+    private final LongList rowValueIsNotNull = new LongList();
+    private final int rootLen;
+    private final FilesFacade ff;
+    private final MemoryMAR symbolMapMem = Vm.getMARInstance();
+    private final int mkDirMode;
+    private final String tableName;
+    private final String walName;
+    private final int walId;
+    private final SequencerMetadata metadata;
+    private final WalWriterEvents events;
+    private final TableSequencerAPI tableSequencerAPI;
+    private final CairoConfiguration configuration;
+    private final ObjList<Runnable> nullSetters;
+    private final RowImpl row = new RowImpl();
+    private final TableWriterSPI walMetadataUpdater = new WalMetadataUpdaterBackend();
+    private final AlterOperationValidationBackend alterOperationValidationBackend = new AlterOperationValidationBackend();
+    private long walLockFd = -1;
+    private long segmentLockFd = -1;
+    private int columnCount;
+    private long currentTxnStartRowNum = -1;
+    private long rowCount = -1;
+    private int segmentId = -1;
+    private long segmentStartMillis;
+    private boolean txnOutOfOrder = false;
+    private long txnMinTimestamp = Long.MAX_VALUE;
+    private long txnMaxTimestamp = -1;
+    private boolean rollSegmentOnNextRow = false;
+    private WalWriterRollStrategy rollStrategy = new WalWriterRollStrategy() {
```

Comment:
nit: could be moved to a no-op static field to avoid an extra allocation.

Reply:
i think this line goes and the `WalWriterRollStrategy` class goes to the bin when we merge Adam's change.
maybe we could leave it as it is until then?

Reply:
Makes sense. Resolving this one.